{influxdb, influxdb2}: update to fix build on Rust 1.78

magneticflux- opened the pull request:
**This also fixes a new build failure after Rust was updated.**

New patch addresses issues with Rust 1.78 until `influxdb2` updates `flux >= v0.195.0`. Removes `no-deny-warnings.patch` and previous Rust version patch as they are not needed anymore.

Fixes #319973

## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

- v2.7.2 does not exist. We do not speak of v2.7.2. 
- v2.7.3 updates Flux to 0.194.3
- https://github.com/influxdata/influxdb/releases/tag/v2.7.4
- https://github.com/influxdata/influxdb/releases/tag/v2.7.5
- https://github.com/influxdata/influxdb/releases/tag/v2.7.6

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

magneticflux- committed:
influxdb2: 2.7.1 -> 2.7.6

New patch addresses issues with Rust 1.78 until `influxdb2` updates `flux >= v0.195.0`. Removes `no-deny-warnings.patch` and previous Rust version patch as they are not needed anymore.

Co-authored-by: éclairevoyant <848000+eclairevoyant@users.noreply.github.com>

magneticflux- commented:
Result of `nixpkgs-review pr 319969` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>influxdb2</li>
    <li>influxdb2-server</li>
  </ul>
</details>

eclairevoyant commented on pkgs/servers/nosql/influxdb2/default.nix:
```diff
@@ -36,21 +36,20 @@ let
-      # Fix build with recent rust versions
+      # Fix build on Rust 1.78 (included after v0.195.0)
       (fetchpatch {
-        url = "https://github.com/influxdata/flux/commit/6dc8054cfeec4b65b5c7ae786d633240868b8589.patch";
+        url = "https://github.com/influxdata/flux/pull/5484.patch";
```
commit hashes are preferred to PR #s for stability
```suggestion
        url = "https://github.com/influxdata/flux/commit/68c831c40b396f0274f6a9f97d77707c39970b02.patch";
```

Scrumplex replied:
I think this patch should get a `name` so it's easier to tell what it's about

eclairevoyant commented on pkgs/servers/nosql/influxdb2/default.nix:
```diff
@@ -36,21 +36,20 @@ let
     ];
     sourceRoot = "${src.name}/libflux";
-    cargoSha256 = "sha256-MoI5nxLGA/3pduZ+vgmSG3lm3Nx58SP+6WXQl2pX9Lc=";
+    cargoSha256 = "sha256-O+t4f4P5291BuyARH6Xf3LejMFEQEBv+qKtyjHRhclA=";
```
nit
```suggestion
    cargoHash = "sha256-O+t4f4P5291BuyARH6Xf3LejMFEQEBv+qKtyjHRhclA=";
```

eclairevoyant reviewed (commented):
Can we also apply this patch to influxdb, which faces a similar build failure?

Scrumplex commented:
Result of `nixpkgs-review pr 319969` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>influxdb2</li>
    <li>influxdb2-server</li>
  </ul>
</details>

magneticflux- committed:
influxdb: 1.10.5 -> 1.10.7

New patch addresses issues with Rust 1.78 until `influxdb` updates `flux >= v0.195.0`.

magneticflux- commented:
I also included the patch for `influxdb`. I updated to the latest patch version `v1.10.7`, but there is a new minor version `v1.11.5` that I'm not sure if we want to update to yet.

nagisa commented on pkgs/servers/nosql/influxdb2/no-deny-warnings.patch:
```diff
@@ -1,10 +0,0 @@
---- a/flux/src/lib.rs
-+++ b/flux/src/lib.rs
-@@ -1,5 +1,3 @@
--#![cfg_attr(feature = "strict", deny(warnings, missing_docs))]
```
Side question: should/Could we perhaps use `--cap-lints warn` by default for rust crates in nixpkgs?

magneticflux- replied:
That seems like a good idea, would it go in `NIX_RUSTFLAGS` somewhere?

eclairevoyant commented:
@ofborg build scrutiny nixosTests.scrutiny

eclairevoyant commented:
btw, CI indicates that libflux [fails to build on aarch64](https://logs.ofborg.org/?key=nixos/nixpkgs.319969&attempt_id=86a484fb-4534-4b56-b8df-d9427b690357)

magneticflux- committed:
{influxdb, influxdb2}: fix build on aarch64-linux

An inadvertent reliance on `char` being signed crept into `libflux`;
this patch specifies a signed 8-bit type for the literal and then
explicitly casts to the appropriate platform `c_char` type.

magneticflux- commented:
> btw, CI indicates that libflux [fails to build on aarch64](https://logs.ofborg.org/?key=nixos/nixpkgs.319969&attempt_id=86a484fb-4534-4b56-b8df-d9427b690357)

Already on it :laughing: https://github.com/influxdata/flux/issues/5495

eclairevoyant commented:
@ofborg build scrutiny nixosTests.scrutiny

magneticflux- commented:
Luckily it seems my "blind" quick fix was enough to get AArch64 working again, but for the future is there a quick way to run `nixosTests` in a cross-platform VM? It seems like it should be "straightforward" (yeah, right) since the tests already run in a VM.

eclairevoyant commented:
building tests requires building the package, which requires one of:
* native platform support
* emulation (binfmt or the like), which is far slower inherently
* cross-compilation, which requires building the `pkgsCross` toolchains and may be slower indirectly

magneticflux- commented:
In case this helps anyone else, this is how I ran `nixosTests` under binfmt emulation:
- Add `boot.binfmt.emulatedSystems = [ "aarch64-linux" ];` to my remote server
- Add `nix.settings.buildMachines.${myserver}.systems = [ "x86_64-linux" "i686-linux" "aarch64-linux" ];` to my local machine
- Run `{nix,nom}-build -A pkgsCross.aarch64-multiplatform.nixosTests.influxdb2` on my local machine

éclair is right, when it has to rely on binfmt emulation it's _very_ slow (10x slower?), which hurts the most on single-threaded operations. The Rust builds saturate my server's 80 threads at first, but there's a long tail that takes a while to complete.

eclairevoyant commented:
If you used `pkgsCross`, that's cross-comp, not emulation; your slowdown might be due to `pkgsCross` not being cached by hydra.

Emulated compilation would be something like `nix build -f . influxdb2 --system aarch64-linux`

magneticflux- commented:
As far as I can tell, it downloaded a bunch of required derivations from Hydra and then compiled `influxdb` and friends (maybe `go` itself?) using `binfmt` (because it's not set up for cross compilation? idk). I had to run it with `-j 0` to prevent it from trying to build `nixos-test-driver-influxdb2` on my local x86_64-linux machine and failing because it was using aarch64 Python, but after that it started running the tests on my server, inside QEMU (albeit veeeeeery slowly, taking 40 minutes to boot the VM. It's actually absurd how slow it is lol).
